src/scheduler.rs

```diff
@@ -644,14 +608,13 @@ struct ConnectionState {
 
 impl ConnectionState {
     /// Shutdown this connection completely.
-    async fn stop(&self) {
+    async fn stop(&self) -> Result<()> {
         // Trigger shutdown of the run loop.
         self.stop_sender
             .send(())
             .await
-            .expect("stop, missing receiver");
-        // Wait for a notification that the run loop has been shutdown.
-        self.shutdown_receiver.recv().await.ok();
+            .context("failed to stop, missing receiver")?;
```

Comment:
I'm wondering whether we still need the stop_senders though. Can't we just `cancel()` the task like we do for the ephemeral task?

I had assumed that we have this stop_sender and shutdown_receiver stuff so that we can do as much work as possible in `pre_stop()`. @dignifiedquire is there another reason, or can we just stop the loops using `handle.cancel().await;`?

Reply:
> I'm wondering whether we still need the stop_senders though. Can't we just cancel() the task like we do for the ephemeral task?

Ideally we should use stop channels in all loops: try to read from them at the end of every loop iteration and exit cleanly if stop signal was received. Currently loop is just raced against the future receiving the stop signal, which is not better than `cancel()`. This may stop the loop anywhere, even in the middle of some non-atomic operation, e.g. after creating the group in `dc_receive_imf` but before adding a message there.